### PR TITLE
new project psd option

### PR DESCRIPTION
**Pull request by Huangzizhou**
# Description

A new option of projecting the hessian matrix following this [repo](https://github.com/honglin-c/abs-psd), instead of clamping the negative eigenvalues to zero, it flips the sign of negative eigenvalues. 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] Unit test "Project to PSD"

### CODE REVIEW

File: src/ipc/utils/eigen_ext.hpp [hunk 1]
```diff
@@ -138,6 +138,8 @@ project_to_pd(
     const Eigen::Matrix<_Scalar, _Rows, _Cols, _Options, _MaxRows, _MaxCols>& A,
     double eps = 1e-8);
 
+enum class ProjectType { None, Clamp, Abs };
```
Comment (zfergus) [inline src/ipc/utils/eigen_ext.hpp:141]: Can you rename to `PSDProjectionMethod` to make it more self explainable?

Also, can you use all caps for enum members to match the repo style?

File: src/ipc/utils/eigen_ext.tpp [hunk 2]
```diff
@@ -84,12 +87,27 @@ project_to_psd(
     }
     Eigen::DiagonalMatrix<double, Eigen::Dynamic> D(eigensolver.eigenvalues());
     // Save a little time and only project the negative values
-    for (int i = 0; i < A.rows(); i++) {
-        if (D.diagonal()[i] < 0.0) {
-            D.diagonal()[i] = 0.0;
-        } else {
-            break;
+    switch (type)
```
Comment (zfergus) [inline src/ipc/utils/eigen_ext.tpp:90]: Can you place the switch inside the for loop to reduce code redundancy? Branch prediction should mitigate the performance difference.

File: src/ipc/utils/eigen_ext.tpp [hunk 3]
```diff
@@ -84,12 +87,27 @@ project_to_psd(
...
+            } else {
+                break;
+            }
         }
+        break;
+    }
+    case ProjectType::Abs:
+    {
+        for (int i = 0; i < A.rows(); i++)
+            D.diagonal()[i] = std::abs(D.diagonal()[i]);
```
Comment (zfergus) [inline src/ipc/utils/eigen_ext.tpp:106]: The eigen values are sorted, so you can break once you encounter the first non-negative value.

Also, can you use braces here to match the repo style?

**Review by zfergus — CHANGES_REQUESTED** [review]
These changes look great! I left a couple of minor comments. I'm happy to merge once they are resolved.

**Comment by zfergus** [comment]
Also, it looks like you need to update the Python bindings. LMK if you need any help with that.

File: CMakeLists.txt [hunk 4]
```diff
@@ -62,7 +62,7 @@ project(IPCToolkit
         VERSION "1.3.0")
 
 option(IPC_TOOLKIT_BUILD_TESTS                "Build unit-tests"  ${IPC_TOOLKIT_TOPLEVEL_PROJECT})
-option(IPC_TOOLKIT_BUILD_PYTHON               "Build Python bindings"                         OFF)
+option(IPC_TOOLKIT_BUILD_PYTHON               "Build Python bindings"                          ON)
```
Comment (zfergus) [inline CMakeLists.txt:65]: Why change this?

Comment (Huangzizhou) [reply CMakeLists.txt:65]: oh sorry that was for testing...

**Review by zfergus — APPROVED** [review]

**Pull request merged**